### PR TITLE
Fix non-threadsafe creation of adapter for type with cyclic dependency

### LINKED ISSUE #625: FutureTypeAdapter.read() throws unhelpful exception
```
What steps will reproduce the problem?
When FutureTypeAdapter's delegate is null, the read() method throws an 
IllegalStateException with no message.  This then gets turned into a 
JsonSyntaxException

These two exceptions don't make it easy to diagnose what went wrong, and you 
can't easily tell if the issue is caused by a programming error, a bug in Gson, 
or with the json being parsed.  There's also no indication of what the 
developer can do to avoid hitting this exception. Adding javasdoc and a helpful 
method to the exception woudl be very nice. 

Below is a sample stack trace. 

E/Timber  ( 3055): Caused by: retrofit.converter.ConversionException: 
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException
E/Timber  ( 3055):  at 
retrofit.converter.GsonConverter.fromBody(GsonConverter.java:67)
E/Timber  ( 3055):  at 
retrofit.RestAdapter$RestHandler.invokeRequest(RestAdapter.java:362)
E/Timber  ( 3055):  ... 9 more
E/Timber  ( 3055): Caused by: com.google.gson.JsonSyntaxException: 
java.lang.IllegalStateException
E/Timber  ( 3055):  at 
com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(Reflecti
veTypeAdapterFactory.java:187)
E/Timber  ( 3055):  at com.google.gson.Gson.fromJson(Gson.java:805)
E/Timber  ( 3055):  at com.google.gson.Gson.fromJson(Gson.java:770)
E/Timber  ( 3055):  at 
retrofit.converter.GsonConverter.fromBody(GsonConverter.java:63)
E/Timber  ( 3055):  ... 10 more
E/Timber  ( 3055): Caused by: java.lang.IllegalStateException
E/Timber  ( 3055):  at 
com.google.gson.Gson$FutureTypeAdapter.read(Gson.java:885)
E/Timber  ( 3055):  at 
com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRunt
imeTypeWrapper.java:40)
E/Timber  ( 3055):  at 
com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(Collecti
onTypeAdapterFactory.java:81)
E/Timber  ( 3055):  at 
com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(Collecti
onTypeAdapterFactory.java:60)
E/Timber  ( 3055):  at 
com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveType
AdapterFactory.java:95)
E/Timber  ( 3055):  at 
com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(Reflecti
veTypeAdapterFactory.java:183)
E/Timber  ( 3055):  ... 13 more

```

Original issue reported on code.google.com by `MhaleK...@gmail.com` on 29 Jan 2015 at 5:26

### PR DESCRIPTION
**Pull request by Marcono1234**
Fixes #625

https://github.com/google/gson/issues/625#issuecomment-321373437 and https://github.com/google/gson/issues/764#issuecomment-169515991 describe the issue very well.

----

The author of #1830 was slightly faster in providing a fix for this issue, though this pull request here also improves the exception messages.
There is also #767 which might fix this issue as well.

### CODE REVIEW

**Review by atlslscsrv-app — APPROVED** [review]

**Review by atlslscsrv-app — APPROVED** [review]

**Comment by byencho** [comment]
Any plan to merge this? We would really like to see this fixed.

File: gson/src/main/java/com/google/gson/Gson.java [hunk 1]
```diff
@@ -523,46 +524,76 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
...
+            // Publish resolved adapters to all threads
+            // Can only do this for the initial request because cyclic dependency TypeA -> TypeB -> TypeA
+            // would otherwise publish adapter for TypeB which uses not yet resolved adapter for TypeA
+            // See https://github.com/google/gson/issues/625
+            for (Map.Entry<TypeToken<?>, TypeAdapter<?>> resolvedAdapterEntry : threadCalls.entrySet()) {
+              typeTokenCache.putIfAbsent(resolvedAdapterEntry.getKey(), resolvedAdapterEntry.getValue());
+            }
+
+            @SuppressWarnings("unchecked")
+            TypeAdapter<T> actualAdapter = (TypeAdapter<T>) typeTokenCache.get(type);
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/Gson.java:565]: This separate `get` call is needed to preserve the behavior of #2153. It retrieves the adapter `putIfAbsent` added above (since it is also called with `candidate` as value), or the existing one in case another thread concurrently put the adapter into the map.

Note that this cannot guarantee that for nested adapters the same adapter instance is used as well, e.g.:
```
A
  X
B
  X
```
Here when requesting adapters for `A` and `B` concurrently they might refer to different instances of the adapter for `X`. Though that is probably acceptable.

File: gson/src/main/java/com/google/gson/Gson.java [hunk 2]
```diff
@@ -1329,9 +1361,20 @@ public void setDelegate(TypeAdapter<T> typeAdapter) {
...
 
+    public void markBroken() {
+      isBroken = true;
+    }
+
     private TypeAdapter<T> delegate() {
+      TypeAdapter<T> delegate = this.delegate;
+      if (isBroken) {
+        throw new IllegalStateException("Broken adapter has been leaked by TypeAdapterFactory");
+      }
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/Gson.java:1372]: See the corresponding test which shows a situation where this can occur.

File: gson/src/test/java/com/google/gson/GsonTest.java [hunk 3]
```diff
@@ -368,4 +360,223 @@ public CustomClass3() {
...
+
+  /**
+   * Verifies that two threads calling {@link Gson#getAdapter(TypeToken)} do not see the
+   * same unresolved {@link FutureTypeAdapter} instance, since that would not be thread-safe.
+   *
+   * This test constructs the cyclic dependency CustomClassA -> CustomClassB1 -> CustomClassA
+   * and lets one thread wait after the adapter for CustomClassB1 has been obtained (which still
+   * contains the nested unresolved FutureTypeAdapter for CustomClassA).
+   */
+  public void testGetAdapterFutureAdapterConcurrency() throws Exception {
```
Comment (Marcono1234) [inline gson/src/test/java/com/google/gson/GsonTest.java:465]: This test is quite verbose but it should hopefully cover exactly the situation #625 is about; reverting the `Gson` changes of this pull request but keeping these `GsonTest` changes triggers the bug described in that issue (`IllegalStateException` being thrown).

**Review by Marcono1234 — COMMENTED** [review]
@eamonnmcmanus could you please have a look?

Unfortunately these changes are a bit complex (both in `Gson` and `GsonTest`). A simpler solution would be possible see #767 or #1830, at the cost of producing less helpful exception messages in some cases, or not preventing broken adapters when `TypeAdapterFactory` implementations discard exceptions thrown by `getAdapter` and fall back to requesting the adapter for a different type (note that the tests here cover that scenario).

File: gson/src/main/java/com/google/gson/Gson.java [hunk 4]
```diff
@@ -523,46 +524,76 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
...
-
           call.setDelegate(candidate);
+          // Replace future adapter with actual adapter
+          threadCalls.put(type, candidate);
+
+          if (isInitialAdapterRequest) {
+            // Publish resolved adapters to all threads
+            // Can only do this for the initial request because cyclic dependency TypeA -> TypeB -> TypeA
+            // would otherwise publish adapter for TypeB which uses not yet resolved adapter for TypeA
+            // See https://github.com/google/gson/issues/625
```
Comment (eamonnmcmanus) [inline gson/src/main/java/com/google/gson/Gson.java:559]: Just so I understand, in a situation like this the second `TypeA` will get a `FutureTypeAdapter`, `TypeB` will reference that `FutureTypeAdapter`, then the first `TypeA` will get the actual `TypeAdapter` returned by the `TypeAdapterFactory`. `TypeB` will continue to reference the `FutureTypeAdapter` it got for `TypeA`, but its delegate will have been updated to be the actual `TypeAdapter`. Is that right?

**Review by eamonnmcmanus — CHANGES_REQUESTED** [review]
Thanks for all the work that has gone into this!

I'm concerned that the logic here is quite hard to understand, and may be trying to solve problems that we could get away with not solving. Perhaps we can simplify, while still fixing #625?

File: gson/src/main/java/com/google/gson/Gson.java [hunk 5]
```diff
@@ -161,8 +162,8 @@ public final class Gson {
    * lookup would stack overflow. We cheat by returning a proxy type adapter.
    * The proxy is wired up once the initial adapter has been created.
    */
-  private final ThreadLocal<Map<TypeToken<?>, FutureTypeAdapter<?>>> calls
-      = new ThreadLocal<>();
+  // Uses LinkedHashMap because iteration order is important, see getAdapter() implementation below
```
[the remark this answers is not in the document]
Comment (Marcono1234) [reply gson/src/main/java/com/google/gson/Gson.java:165]: Have reverted usage of `LinkedHashMap` as part of simplifying this pull request. I hope the newly adjusted doc comment is fine.

File: gson/src/main/java/com/google/gson/Gson.java (hunk 4 above)
Comment (Marcono1234) [reply gson/src/main/java/com/google/gson/Gson.java:559]: Yes that is correct. And in the past the issue was that the adapter for `TypeB` (which references the `FutureTypeAdapter`) was already published to other threads before the `FutureTypeAdapter` had been resolved. That was not thread-safe.

File: gson/src/main/java/com/google/gson/Gson.java (hunk 1 above)
Comment (Marcono1234) [reply gson/src/main/java/com/google/gson/Gson.java:565]: That is fine for me. When I created #2153 (which introduced this `putIfAbsent` functionality) I thought it might be nice to have, but as far as I know there was no issue explicitly requesting this.

File: gson/src/main/java/com/google/gson/Gson.java [hunk 6]
```diff
@@ -523,46 +524,76 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
...
     } finally {
-      threadCalls.remove(type);
-
-      if (requiresThreadLocalCleanup) {
+      if (isInitialAdapterRequest) {
         calls.remove();
       }
+      if (!foundCandidate) {
+        Iterator<TypeAdapter<?>> adaptersIterator = threadCalls.values().iterator();
+        // Skip existing non-broken adapters
```
[the remark this answers is not in the document]
Comment (Marcono1234) [reply gson/src/main/java/com/google/gson/Gson.java:581]: Yes, I have now removed this logic here again because this is indeed likely a contrived corner case, and the current implementation does not handle that in a better way either.

File: gson/src/test/java/com/google/gson/GsonTest.java [hunk 7]
```diff
@@ -368,4 +360,223 @@ public CustomClass3() {
...
     }
   }
+
+  /**
+   * Verifies that {@link Gson#getAdapter(TypeToken)} does not put broken adapters
+   * into {@code typeTokenCache} when caller of nested {@code getAdapter} discards
+   * exception, e.g.:
+   *
+   * Field dependencies:
+   * ClassA
```
[the remark this answers is not in the document]
Comment (Marcono1234) [reply gson/src/test/java/com/google/gson/GsonTest.java:370]: Have removed this test again since it was only relevant for the contrived situation of discarding exceptions from `getAdapter`, which I have removed as well.

File: gson/src/main/java/com/google/gson/Gson.java [hunk 8]
```diff
@@ -523,47 +532,55 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
...
-    }
-
-    // the key and value type parameters always agree
-    @SuppressWarnings("unchecked")
-    FutureTypeAdapter<T> ongoingCall = (FutureTypeAdapter<T>) threadCalls.get(type);
-    if (ongoingCall != null) {
-      return ongoingCall;
+      threadLocalAdapterResults.set(threadCalls);
+      isInitialAdapterRequest = true;
+    } else {
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/Gson.java:541]: Have moved the code here in the `else` block; otherwise the `ongoingCall` check will be unsuccessful since the `if` statement just created an empty map as `threadCalls` value.

File: gson/src/main/java/com/google/gson/Gson.java [hunk 9]
```diff
@@ -523,47 +532,55 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
...
-          }
-
           call.setDelegate(candidate);
-          return candidate;
+          // Replace future adapter with actual adapter
+          threadCalls.put(type, candidate);
+          break;
         }
       }
-      throw new IllegalArgumentException("GSON (" + GsonBuildConfig.VERSION + ") cannot handle " + type);
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/Gson.java:559]: Have moved this and the `return` outside the `try`-`finally`.

**Review by Marcono1234 — COMMENTED** [review]
Thanks for the review! I have simplified this pull request now as discussed in the above review comments:
- omitting the handling for the contrived case "exception from `getAdapter` being discarded"
- omitting preserving existsing adapter instance for concurrent `getAdapter` calls, since that behavior could not be guaranteed in all cases anymore

**Comment by eamonnmcmanus** [comment]
Thanks, this looks great!

**Comment by richardm-stripe** [comment]
Hello @eamonnmcmanus! Will this be in GSON 2.11 and can you share when 2.11 is expected to land? Our users run into this sometimes, and we'd like to judge whether to wait for 2.11 or put a workaround for this into our library.

Thank you!

**Comment by byencho** [comment]
Thanks @eamonnmcmanus !
